In [ ]:
from sklearn.neural_network import MLPClassifier
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

Integrantes:

- Jhoan Sebastian Andica 1859966-3743
- Diego Fernando Delgado 1859819-3743
- David Rios Gonzales 1765662-3743

In [ ]:
!wget https://raw.githubusercontent.com/Sebastian1811/DataSetTallerRedesNeuronales/main/nasa.csv

--2022-10-15 14:26:20--  https://raw.githubusercontent.com/Sebastian1811/DataSetTallerRedesNeuronales/main/nasa.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.108.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2425254 (2.3M) [text/plain]
Saving to: ‘nasa.csv’

nasa.csv            100%[===================>]   2.31M  --.-KB/s    in 0.07s   

2022-10-15 14:26:20 (31.8 MB/s) - ‘nasa.csv’ saved [2425254/2425254]



In [ ]:
data = pd.read_csv("nasa.csv")

In [ ]:
data.head()

Neo Reference ID     Name  Absolute Magnitude  Est Dia in KM(min)  \
0           3703080  3703080                21.6            0.127220   
1           3723955  3723955                21.3            0.146068   
2           2446862  2446862                20.3            0.231502   
3           3092506  3092506                27.4            0.008801   
4           3514799  3514799                21.6            0.127220   

   Est Dia in KM(max)  Est Dia in M(min)  Est Dia in M(max)  \
0            0.284472         127.219879         284.472297   
1            0.326618         146.067964         326.617897   
2            0.517654         231.502122         517.654482   
3            0.019681           8.801465          19.680675   
4            0.284472         127.219879         284.472297   

   Est Dia in Miles(min)  Est Dia in Miles(max)  Est Dia in Feet(min)  ...  \
0               0.079051               0.176763            417.388066  ...   
1               0.090762               0.202951            479.225620  ...   
2               0.143849               0.321655            759.521423  ...   
3               0.005469               0.012229             28.876199  ...   
4               0.079051               0.176763            417.388066  ...   

   Asc Node Longitude Orbital Period  Perihelion Distance  Perihelion Arg  \
0          314.373913     609.599786             0.808259       57.257470   
1          136.717242     425.869294             0.718200      313.091975   
2          259.475979     643.580228             0.950791      248.415038   
3           57.173266     514.082140             0.983902       18.707701   
4           84.629307     495.597821             0.967687      158.263596   

   Aphelion Dist  Perihelion Time  Mean Anomaly  Mean Motion  Equinox  \
0       2.005764     2.458162e+06    264.837533     0.590551    J2000   
1       1.497352     2.457795e+06    173.741112     0.845330    J2000   
2       1.966857     2.458120e+06    292.893654     0.559371    J2000   
3       1.527904     2.457902e+06     68.741007     0.700277    J2000   
4       1.483543     2.457814e+06    135.142133     0.726395    J2000   

   Hazardous  
0       True  
1      False  
2       True  
3      False  
4       True  

[5 rows x 40 columns]

In [ ]:
data.describe()

Neo Reference ID          Name  Absolute Magnitude  Est Dia in KM(min)  \
count      4.687000e+03  4.687000e+03         4687.000000         4687.000000   
mean       3.272298e+06  3.272298e+06           22.267865            0.204604   
std        5.486011e+05  5.486011e+05            2.890972            0.369573   
min        2.000433e+06  2.000433e+06           11.160000            0.001011   
25%        3.097594e+06  3.097594e+06           20.100000            0.033462   
50%        3.514799e+06  3.514799e+06           21.900000            0.110804   
75%        3.690060e+06  3.690060e+06           24.500000            0.253837   
max        3.781897e+06  3.781897e+06           32.100000           15.579552   

       Est Dia in KM(max)  Est Dia in M(min)  Est Dia in M(max)  \
count         4687.000000        4687.000000        4687.000000   
mean             0.457509         204.604203         457.508906   
std              0.826391         369.573402         826.391249   
min              0.002260           1.010543           2.259644   
25%              0.074824          33.462237          74.823838   
50%              0.247765         110.803882         247.765013   
75%              0.567597         253.837029         567.596853   
max             34.836938       15579.552413       34836.938254   

       Est Dia in Miles(min)  Est Dia in Miles(max)  Est Dia in Feet(min)  \
count            4687.000000            4687.000000           4687.000000   
mean                0.127135               0.284283            671.273653   
std                 0.229642               0.513496           1212.511199   
min                 0.000628               0.001404              3.315431   
25%                 0.020792               0.046493            109.784247   
50%                 0.068850               0.153954            363.529809   
75%                 0.157727               0.352688            832.798679   
max                 9.680682              21.646663          51114.018738   

       ...  Semi Major Axis  Inclination  Asc Node Longitude  Orbital Period  \
count  ...      4687.000000  4687.000000         4687.000000     4687.000000   
mean   ...         1.400264    13.373844          172.157275      635.582076   
std    ...         0.524154    10.936227          103.276777      370.954727   
min    ...         0.615920     0.014513            0.001941      176.557161   
25%    ...         1.000635     4.962341           83.081208      365.605031   
50%    ...         1.240981    10.311836          172.625393      504.947292   
75%    ...         1.678364    19.511681          255.026909      794.195972   
max    ...         5.072008    75.406667          359.905890     4172.231343   

       Perihelion Distance  Perihelion Arg  Aphelion Dist  Perihelion Time  \
count          4687.000000     4687.000000    4687.000000     4.687000e+03   
mean              0.813383      183.932151       1.987144     2.457728e+06   
std               0.242059      103.513035       0.951519     9.442264e+02   
min               0.080744        0.006918       0.803765     2.450100e+06   
25%               0.630834       95.625916       1.266059     2.457815e+06   
50%               0.833153      189.761641       1.618195     2.457973e+06   
75%               0.997227      271.777557       2.451171     2.458108e+06   
max               1.299832      359.993098       8.983852     2.458839e+06   

       Mean Anomaly  Mean Motion  
count   4687.000000  4687.000000  
mean     181.167927     0.738242  
std      107.501623     0.342627  
min        0.003191     0.086285  
25%       87.006918     0.453289  
50%      185.718889     0.712946  
75%      276.531946     0.984669  
max      359.917991     2.039000  

[8 rows x 35 columns]

Debido a la desproporción de las filas con respecto al atributo de "Hazardous" (Riesgoso), se crearon dos nuevos dataframes a partir del original, donde todas las filas con atributo "True" se eligieron y de aquellas filas que tuvieran como atributo "False", se escogieron el 20% de forma aleatoria de dichas filas.

In [ ]:
new_data_for_true = data.where(data["Hazardous"] == 1)
new_data_for_true.dropna().shape[0]

755

In [ ]:
new_data_for_false = data.where(data["Hazardous"] == 0)
new_data_for_false.dropna().shape[0]

3932

In [ ]:
final_data = pd.concat([new_data_for_true.dropna(), new_data_for_false.dropna().sample(frac=0.2, replace=True, random_state=1)])

Se escogieron 9 atributos de entradad para el entrenamiento de la red que se consideraran relevantes y a su vez no tuvieran una correlación entre ellos. Para los atributos de salida, se escogió el atributo de "Peligrosidad".

In [ ]:
x = final_data[["Absolute Magnitude","Est Dia in KM(min)","Est Dia in KM(max)","Relative Velocity km per hr","Miss Dist.(kilometers)","Inclination","Orbital Period", "Perihelion Distance","Aphelion Dist"]]
y = final_data[["Hazardous"]]

In [ ]:
y.head()

Hazardous
0       True
2       True
4       True
9       True
22      True

Se convierten los atributos "True" y "False" a notación binaria.

In [ ]:
yf = y.astype(int)

In [ ]:
yf

Hazardous
0             1
2             1
4             1
9             1
22            1
...         ...
274           0
1906          0
1558          0
1822          0
3218          0

[1541 rows x 1 columns]

In [ ]:
y.value_counts()

Hazardous
False        786
True         755
dtype: int64

In [ ]:
print(x.describe())

       Absolute Magnitude  Est Dia in KM(min)  Est Dia in KM(max)  \
count         1541.000000         1541.000000         1541.000000   
mean            21.487637            0.235409            0.526390   
std              2.574022            0.269520            0.602665   
min             14.400000            0.001839            0.004112   
25%             19.700000            0.069913            0.156329   
50%             21.100000            0.160160            0.358129   
75%             22.900000            0.305179            0.682402   
max             30.800000            3.503926            7.835018   

       Relative Velocity km per hr  Miss Dist.(kilometers)  Inclination  \
count                  1541.000000            1.541000e+03  1541.000000   
mean                  54706.782644            3.929877e+07    13.325604   
std                   27037.548636            2.182943e+07    10.865139   
min                    5028.027688            2.660989e+04     0.118339   
25%

In [ ]:
x.count()

Absolute Magnitude             1541
Est Dia in KM(min)             1541
Est Dia in KM(max)             1541
Relative Velocity km per hr    1541
Miss Dist.(kilometers)         1541
Inclination                    1541
Orbital Period                 1541
Perihelion Distance            1541
Aphelion Dist                  1541
dtype: int64

In [ ]:
x.count(numeric_only=True)

Absolute Magnitude             1541
Est Dia in KM(min)             1541
Est Dia in KM(max)             1541
Relative Velocity km per hr    1541
Miss Dist.(kilometers)         1541
Inclination                    1541
Orbital Period                 1541
Perihelion Distance            1541
Aphelion Dist                  1541
dtype: int64

Se hace el escalamiento de los datos debido a que se usa la función sigmoide como activación, entonces los datos se transforman a una distribución entre [-1, 1].

In [ ]:
from sklearn.preprocessing import StandardScaler
scal = StandardScaler()
scal.fit(x) #Entrenamiento
xf = scal.transform(x) #Genera el nuevo dataset

In [ ]:
xf

array([[ 0.04366702, -0.40154423, -0.40154423, ..., -0.03863276,
         0.18913847,  0.00306878],
       [-0.46154314, -0.01450025, -0.01450025, ...,  0.05636031,
         0.79671617, -0.03823358],
       [ 0.04366702, -0.40154423, -0.40154423, ..., -0.35732775,
         0.86873815, -0.55130329],
       ...,
       [ 1.48157287, -0.78779923, -0.78779923, ...,  0.12203437,
         1.35995863, -0.10357789],
       [-0.22836922, -0.22193549, -0.22193549, ...,  0.22991515,
         1.56363721, -0.03310852],
       [ 0.28966551, -0.52094055, -0.52094055, ...,  2.59653495,
         2.27767834,  2.06604833]])

Se elige un 20% de los datos para testeo y el 80% restante para entrenamiento.

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    xf, yf, test_size=0.2, random_state=35)

In [ ]:
print(X_train.shape, X_test.shape)
print(y_train.shape, y_test.shape)

(1232, 9) (309, 9)
(1232, 1) (309, 1)


MODELO KOLMOGOROV DE 2 CAPAS

In [ ]:
modelo2capas = MLPClassifier(
    hidden_layer_sizes=(9,17),
    activation="logistic",
    solver="lbfgs",
    verbose=True,
    max_iter=2000
)

In [ ]:
modelo2capas.fit(X_train,y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:549: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


MLPClassifier(activation='logistic', hidden_layer_sizes=(9, 17), max_iter=2000,
              solver='lbfgs', verbose=True)

In [ ]:
modelo2capas.score(X_test,y_test)

0.7961165048543689

In [ ]:
modelo2capas.loss_

0.13300063616096086

In [ ]:
y_pred = modelo2capas.predict(X_test)

In [ ]:
from sklearn.metrics import classification_report,confusion_matrix

print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.75      0.86      0.80       147
           1       0.86      0.73      0.79       162

    accuracy                           0.80       309
   macro avg       0.80      0.80      0.80       309
weighted avg       0.80      0.80      0.80       309



Según al reporte de clasificación:

- Hay un 80% que el modelo puede predecir correctamente los datos como Peligrosos
- Hay un 80% de los datos predecidos que sí son catalogados como Peligrosos
- En cuanto al rendimiento del modelo, se obtuvo un 80%, el cual, tomando en cuenta un umbral del 75% de aceptación, se considera el modelo como correcto.

In [ ]:
print(confusion_matrix(y_test,y_pred))

[[127  20]
 [ 43 119]]


MODELO KOLOMOGOROV DE 3 CAPAS

In [ ]:
modelo3capas = MLPClassifier(
    hidden_layer_sizes=(9,17,3),
    activation="logistic",
    solver="lbfgs",
    verbose=True,
    max_iter=2000
)

In [ ]:
modelo3capas.fit(X_train,y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:549: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


MLPClassifier(activation='logistic', hidden_layer_sizes=(9, 17, 3),
              max_iter=2000, solver='lbfgs', verbose=True)

In [ ]:
modelo3capas.score(X_test,y_test)

0.8511326860841424

In [ ]:
modelo3capas.loss_

0.11032731180678597

In [ ]:
y_pred = modelo3capas.predict(X_test)

In [ ]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.84      0.84      0.84       147
           1       0.86      0.86      0.86       162

    accuracy                           0.85       309
   macro avg       0.85      0.85      0.85       309
weighted avg       0.85      0.85      0.85       309



In [ ]:
print(confusion_matrix(y_test,y_pred))

[[124  23]
 [ 23 139]]


Según al reporte de clasificación:
  - Hay un 85% que el modelo puede predecir correctamente los datos como Peligrosos
  - Hay un 85% de los datos predecidos que sí son catalogados como Peligrosos
  - En cuanto al rendimiento del modelo, se obtuvo un 85%, el cual, tomando en cuenta un umbral del 75% de aceptación, se considera el modelo como correcto.